File: src/dcal_annotations/validation.py
from __future__ import annotations

import hashlib
import json
import re
from collections import Counter
from dataclasses import dataclass
from typing import Any, Iterable

from .taxonomy import Taxonomy


SHA256_RE = re.compile(r"^[0-9a-f]{64}$")
FIELD_CODE_RE = re.compile(r"^[a-z][a-z0-9_]*$")
ANNOTATION_SCHEMA_VERSION = "dcal.annotation.v1"
GOLD_SCHEMA_VERSION = "dcal.gold.v1"


@dataclass(frozen=True)
class ValidationIssue:
    path: str
    message: str

    def __str__(self) -> str:
        return f"{self.path}: {self.message}"


class ExportValidationError(ValueError):
    def __init__(self, issues: Iterable[ValidationIssue]):
        self.issues = tuple(issues)
        super().__init__("\n".join(str(issue) for issue in self.issues))


def _require_string(value: Any, path: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{path} must be a non-empty string")
    return value


def _require_positive_int(value: Any, path: str) -> int:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError(f"{path} must be a positive integer")
    if int(value) != value or value < 1:
        raise ValueError(f"{path} must be a positive integer")
    return int(value)


def _as_tasks(payload: Any) -> list[dict[str, Any]]:
    if isinstance(payload, list):
        tasks = payload
    elif isinstance(payload, dict) and isinstance(payload.get("tasks"), list):
        tasks = payload["tasks"]
    else:
        raise ValueError("export root must be a task array or an object with a tasks array")
    if not all(isinstance(task, dict) for task in tasks):
        raise ValueError("every exported task must be a JSON object")
    return tasks


def _select_annotation(task: dict[str, Any]) -> dict[str, Any]:
    annotations = task.get("annotations")
    if not isinstance(annotations, list):
        raise ValueError("annotations must be an array")
    active = [
        annotation
        for annotation in annotations
        if isinstance(annotation, dict)
        and annotation.get("was_cancelled") is not True
        and isinstance(annotation.get("result"), list)
        and annotation["result"]
    ]
    ground_truth = [annotation for annotation in active if annotation.get("ground_truth") is True]
    if len(ground_truth) == 1:
        return ground_truth[0]
    if len(ground_truth) > 1:
        raise ValueError("multiple active annotations are marked as ground truth")
    if len(active) != 1:
        raise ValueError(
            "task must have exactly one active annotation or exactly one explicit ground truth"
        )
    return active[0]


def _items(results: list[dict[str, Any]], from_name: str) -> list[dict[str, Any]]:
    return [item for item in results if item.get("from_name") == from_name]


def _single_choice(
    results: list[dict[str, Any]], from_name: str, *, required: bool
) -> str | None:
    matches = _items(results, from_name)
    if not matches:
        if required:
            raise ValueError(f"missing required {from_name} choice")
        return None
    if len(matches) != 1:
        raise ValueError(f"{from_name} must contain exactly one result item")
    choices = matches[0].get("value", {}).get("choices")
    if not isinstance(choices, list) or len(choices) != 1 or not isinstance(choices[0], str):
        raise ValueError(f"{from_name} must contain exactly one choice")
    return choices[0]


def _multiple_choices(results: list[dict[str, Any]], from_name: str) -> list[str]:
    matches = _items(results, from_name)
    if not matches:
        return []
    if len(matches) != 1:
        raise ValueError(f"{from_name} must contain at most one result item")
    choices = matches[0].get("value", {}).get("choices")
    if not isinstance(choices, list) or not all(isinstance(value, str) for value in choices):
        raise ValueError(f"{from_name} choices must be an array of strings")
    if len(set(choices)) != len(choices):
        raise ValueError(f"{from_name} contains duplicate choices")
    return choices


def _single_text(results: list[dict[str, Any]], from_name: str) -> str | None:
    matches = _items(results, from_name)
    if not matches:
        return None
    if len(matches) != 1:
        raise ValueError(f"{from_name} must contain at most one result item")
    value = matches[0].get("value", {}).get("text")
    if isinstance(value, str):
        return value
    if isinstance(value, list) and len(value) == 1 and isinstance(value[0], str):
        return value[0]
    raise ValueError(f"{from_name} must contain exactly one text value")


def _single_number(results: list[dict[str, Any]], from_name: str) -> int:
    matches = _items(results, from_name)
    if len(matches) != 1:
        raise ValueError(f"{from_name} must contain exactly one result item")
    return _require_positive_int(matches[0].get("value", {}).get("number"), from_name)


def _annotation_actor(annotation: dict[str, Any]) -> int | str | None:
    actor = annotation.get("completed_by")
    if isinstance(actor, dict):
        actor = actor.get("id")
    if isinstance(actor, (int, str)) and not isinstance(actor, bool):
        return actor
    return None


def _validate_geometry(result: dict[str, Any]) -> dict[str, Any]:
    value = result.get("value")
    if not isinstance(value, dict):
        raise ValueError("region geometry value must be an object")

    geometry: dict[str, float] = {}
    for name in ("x", "y", "width", "height"):
        raw = value.get(name)
        if isinstance(raw, bool) or not isinstance(raw, (int, float)):
            raise ValueError(f"region {name} must be numeric")
        geometry[name] = float(raw)

    if not 0 <= geometry["x"] <= 100 or not 0 <= geometry["y"] <= 100:
        raise ValueError("region x and y must be between 0 and 100")
    if not 0 < geometry["width"] <= 100 or not 0 < geometry["height"] <= 100:
        raise ValueError("region width and height must be greater than 0 and at most 100")
    if geometry["x"] + geometry["width"] > 100.001:
        raise ValueError("region extends beyond the right page boundary")
    if geometry["y"] + geometry["height"] > 100.001:
        raise ValueError("region extends beyond the bottom page boundary")

    rotation = value.get("rotation", 0)
    if isinstance(rotation, bool) or not isinstance(rotation, (int, float)):
        raise ValueError("region rotation must be numeric")
    if not -360 <= rotation <= 360:
        raise ValueError("region rotation must be between -360 and 360 degrees")

    original_width = _require_positive_int(result.get("original_width"), "original_width")
    original_height = _require_positive_int(result.get("original_height"), "original_height")
    image_rotation = result.get("image_rotation", 0)
    if isinstance(image_rotation, bool) or not isinstance(image_rotation, (int, float)):
        raise ValueError("image_rotation must be numeric")

    return {
        **geometry,
        "rotation": float(rotation),
        "original_width": original_width,
        "original_height": original_height,
        "image_rotation": float(image_rotation),
    }


def _normalize_region(
    geometry_result: dict[str, Any],
    all_results: list[dict[str, Any]],
    taxonomy: Taxonomy,
) -> dict[str, Any]:
    region_id = _require_string(geometry_result.get("id"), "region id")
    labels = geometry_result.get("value", {}).get("rectanglelabels")
    if not isinstance(labels, list) or len(labels) != 1 or not isinstance(labels[0], str):
        raise ValueError(f"region {region_id} must have exactly one region label")
    label = labels[0]
    if label not in taxonomy.region_labels:
        raise ValueError(f"region {region_id} uses unknown label {label!r}")

    scoped = [item for item in all_results if item.get("id") == region_id]
    legibility = _single_choice(scoped, "legibility", required=True)
    assert legibility is not None
    if legibility not in taxonomy.legibility_states:
        raise ValueError(f"region {region_id} uses unknown legibility {legibility!r}")

    structure_role = _single_choice(scoped, "structure_role", required=False) or "none"
    if structure_role not in taxonomy.structure_roles:
        raise ValueError(f"region {region_id} uses unknown structure role {structure_role!r}")

    semantic_type = _single_choice(scoped, "semantic_region_type", required=False)
    if semantic_type is not None and semantic_type not in taxonomy.semantic_region_types:
        raise ValueError(f"region {region_id} uses unknown semantic type {semantic_type!r}")

    reading_order = _single_number(scoped, "reading_order")
    transcription = _single_text(scoped, "transcription")
    field_code = _single_text(scoped, "field_code")
    if field_code is not None:
        field_code = field_code.strip()
        if not field_code:
            field_code = None
        elif not FIELD_CODE_RE.fullmatch(field_code):
            raise ValueError(
                f"region {region_id} field_code must use lowercase snake_case"
            )

    label_contract = taxonomy.region_labels[label]
    if label_contract.get("textual") is True and legibility == "not_applicable":
        raise ValueError(f"region {region_id} is textual and cannot be not_applicable")
    if legibility in {"illegible", "not_applicable"} and transcription is not None:
        if transcription.strip():
            raise ValueError(
                f"region {region_id} must not contain transcription when {legibility}"
            )
        transcription = None
    if (
        label_contract.get("requires_transcription_when_readable") is True
        and legibility in {"legible", "partially_legible"}
        and (transcription is None or not transcription.strip())
    ):
        raise ValueError(f"region {region_id} requires exact transcription")

    return {
        "region_id": region_id,
        "reading_order": reading_order,
        "label": label,
        "structure_role": structure_role,
        "legibility": legibility,
        "semantic_region_type": semantic_type,
        "field_code": field_code,
        "transcription": transcription,
        "geometry": _validate_geometry(geometry_result),
    }


def _normalize_task(task: dict[str, Any], taxonomy: Taxonomy) -> dict[str, Any]:
    data = task.get("data")
    if not isinstance(data, dict):
        raise ValueError("data must be an object")
    _require_string(data.get("image"), "data.image")
    object_id = _require_string(data.get("source_object_id"), "data.source_object_id")
    source_sha256 = _require_string(data.get("source_sha256"), "data.source_sha256")
    if not SHA256_RE.fullmatch(source_sha256):
        raise ValueError("data.source_sha256 must be 64 lowercase hexadecimal characters")
    patient_group_id = _require_string(
        data.get("patient_group_id"), "data.patient_group_id"
    )
    encounter_group_id = _require_string(
        data.get("encounter_group_id"), "data.encounter_group_id"
    )
    page_index = _require_positive_int(data.get("source_page_index"), "data.source_page_index")
    schema_version = _require_string(
        data.get("annotation_schema_version"), "data.annotation_schema_version"
    )
    if schema_version != ANNOTATION_SCHEMA_VERSION:
        raise ValueError(
            f"unsupported annotation schema {schema_version!r}; expected {ANNOTATION_SCHEMA_VERSION!r}"
        )

    writer_group_ids = data.get("writer_group_ids", [])
    if not isinstance(writer_group_ids, list) or not all(
        isinstance(value, str) and value.strip() for value in writer_group_ids
    ):
        raise ValueError("data.writer_group_ids must be an array of non-empty opaque IDs")
    if len(set(writer_group_ids)) != len(writer_group_ids):
        raise ValueError("data.writer_group_ids contains duplicates")

    annotation = _select_annotation(task)
    raw_results = annotation["result"]
    if not all(isinstance(item, dict) for item in raw_results):
        raise ValueError("annotation results must all be objects")
    results: list[dict[str, Any]] = raw_results

    physical_type = _single_choice(results, "physical_document_type", required=True)
    assert physical_type is not None
    if physical_type not in taxonomy.physical_document_types:
        raise ValueError(f"unknown physical document type {physical_type!r}")

    variant = _single_choice(results, "physical_document_variant", required=False)
    if variant is not None:
        variant_contract = taxonomy.physical_document_variants.get(variant)
        if variant_contract is None:
            raise ValueError(f"unknown physical document variant {variant!r}")
        if variant_contract["physical_document_type"] != physical_type:
            raise ValueError(
                f"variant {variant!r} does not belong to physical type {physical_type!r}"
            )

    quality = _multiple_choices(results, "image_quality")
    unknown_quality = sorted(set(quality) - set(taxonomy.image_quality_flags))
    if unknown_quality:
        raise ValueError(f"unknown image quality flags: {unknown_quality!r}")
    if "clear" in quality and len(quality) > 1:
        raise ValueError("image quality 'clear' cannot be combined with defect flags")

    geometry_results = _items(results, "region_label")
    geometry_ids = {
        item.get("id")
        for item in geometry_results
        if isinstance(item.get("id"), str) and item.get("id")
    }
    region_control_names = {
        "reading_order",
        "legibility",
        "structure_role",
        "semantic_region_type",
        "transcription",
        "field_code",
    }
    orphaned_controls = sorted(
        {
            str(item.get("id"))
            for item in results
            if item.get("from_name") in region_control_names
            and item.get("id") not in geometry_ids
        }
    )
    if orphaned_controls:
        raise ValueError(
            "region controls reference missing geometry IDs: "
            f"{orphaned_controls!r}"
        )

    seen_region_ids: set[str] = set()
    regions: list[dict[str, Any]] = []
    for geometry_result in geometry_results:
        if geometry_result.get("type") != "rectanglelabels":
            raise ValueError("region_label results must use rectanglelabels geometry")
        region_id = _require_string(geometry_result.get("id"), "region id")
        if region_id in seen_region_ids:
            raise ValueError(f"duplicate region geometry id {region_id!r}")
        seen_region_ids.add(region_id)
        regions.append(_normalize_region(geometry_result, results, taxonomy))

    if physical_type not in taxonomy.non_catalog_types and not regions:
        raise ValueError("clinical physical pages require at least one annotated region")

    reading_orders = [region["reading_order"] for region in regions]
    if len(reading_orders) != len(set(reading_orders)):
        raise ValueError("region reading_order values must be unique within a page")
    regions.sort(key=lambda region: region["reading_order"])

    notes = _single_text(results, "annotator_notes")
    if notes is not None and not notes.strip():
        notes = None

    record: dict[str, Any] = {
        "schema_version": GOLD_SCHEMA_VERSION,
        "taxonomy_version": taxonomy.version,
        "source": {
            "object_id": object_id,
            "sha256": source_sha256,
            "patient_group_id": patient_group_id,
            "encounter_group_id": encounter_group_id,
            "writer_group_ids": writer_group_ids,
            "page_index": page_index,
        },
        "annotation": {
            "label_studio_task_id": task.get("id"),
            "label_studio_annotation_id": annotation.get("id"),
            "annotator_id": _annotation_actor(annotation),
            "created_at": annotation.get("created_at"),
            "updated_at": annotation.get("updated_at"),
        },
        "classification": {
            "physical_document_type": physical_type,
            "physical_document_variant": variant,
        },
        "image_quality": sorted(quality),
        "annotator_notes": notes,
        "regions": regions,
    }
    canonical = json.dumps(
        record, ensure_ascii=False, sort_keys=True, separators=(",", ":")
    ).encode("utf-8")
    record["record_sha256"] = hashlib.sha256(canonical).hexdigest()
    return record


def normalize_export(payload: Any, taxonomy: Taxonomy) -> list[dict[str, Any]]:
    try:
        tasks = _as_tasks(payload)
    except ValueError as error:
        raise ExportValidationError([ValidationIssue("export", str(error))]) from error

    issues: list[ValidationIssue] = []
    records: list[dict[str, Any]] = []
    for index, task in enumerate(tasks):
        task_id = task.get("id", "unknown")
        try:
            records.append(_normalize_task(task, taxonomy))
        except (AssertionError, KeyError, TypeError, ValueError) as error:
            issues.append(ValidationIssue(f"tasks[{index}] (id={task_id})", str(error)))

    if issues:
        raise ExportValidationError(issues)

    source_hashes = [record["source"]["sha256"] for record in records]
    duplicates = sorted(
        value for value, count in Counter(source_hashes).items() if count > 1
    )
    if duplicates:
        raise ExportValidationError(
            [
                ValidationIssue(
                    "export",
                    f"contains duplicate source SHA-256 values: {duplicates!r}",
                )
            ]
        )
    return records

File: web/lib/validation.ts
import {
  ANNOTATION_SCHEMA,
  NON_REGION_DOCUMENTS,
  taxonomySets,
} from "@/lib/contracts";
import { HttpError } from "@/lib/http";

const REGION_ID = /^reg_[a-f0-9]{12,32}$/;
const FIELD_CODE = /^[a-z][a-z0-9_]*$/;
const TABLE_MAX_ROWS = 100;
const TABLE_MAX_COLUMNS = 12;
const TABLE_MAX_CELL_LENGTH = 10_000;
const TABLE_MAX_TEXT_LENGTH = 100_000;

function invalid(message: string): never {
  throw new HttpError(400, "invalid_annotation", message);
}

function optionalString(value: unknown, label: string) {
  if (value === null || value === undefined || value === "") return null;
  if (typeof value !== "string") invalid(`${label} must be text.`);
  return value;
}

function normalizeTableData(value: unknown, regionNumber: number) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    invalid(`Region ${regionNumber} needs table data.`);
  }
  const input = value as Record<string, unknown>;
  const rows = input.rows;
  const columns = input.columns;
  const headerRows = input.header_rows ?? 0;
  if (!Number.isInteger(rows) || (rows as number) < 1 || (rows as number) > TABLE_MAX_ROWS) {
    invalid(`Region ${regionNumber} table rows must be between 1 and ${TABLE_MAX_ROWS}.`);
  }
  if (!Number.isInteger(columns) || (columns as number) < 1 || (columns as number) > TABLE_MAX_COLUMNS) {
    invalid(`Region ${regionNumber} table columns must be between 1 and ${TABLE_MAX_COLUMNS}.`);
  }
  if (!Number.isInteger(headerRows) || (headerRows as number) < 0 || (headerRows as number) > (rows as number)) {
    invalid(`Region ${regionNumber} has invalid table header rows.`);
  }

  const columnLabels = input.column_labels;
  if (!Array.isArray(columnLabels) || columnLabels.length !== columns) {
    invalid(`Region ${regionNumber} table column types must match the column count.`);
  }
  const normalizedColumnLabels = columnLabels.map((label, offset) => {
    const contract = typeof label === "string" ? taxonomySets.regionLabels.get(label) : undefined;
    if (!contract || contract.textual !== true) {
      invalid(`Region ${regionNumber} table column ${offset + 1} needs a textual content type.`);
    }
    return label as string;
  });

  const rawCells = input.cells;
  if (!Array.isArray(rawCells) || rawCells.length !== rows) {
    invalid(`Region ${regionNumber} table cells must match the row count.`);
  }
  let totalLength = 0;
  const cells = rawCells.map((rawRow, rowIndex) => {
    if (!Array.isArray(rawRow) || rawRow.length !== columns) {
      invalid(`Region ${regionNumber} table row ${rowIndex + 1} must match the column count.`);
    }
    return rawRow.map((cell, columnIndex) => {
      if (typeof cell !== "string" || cell.length > TABLE_MAX_CELL_LENGTH) {
        invalid(`Region ${regionNumber} table cell ${rowIndex + 1},${columnIndex + 1} is invalid.`);
      }
      totalLength += cell.length;
      if (totalLength > TABLE_MAX_TEXT_LENGTH) {
        invalid(`Region ${regionNumber} table text is too large.`);
      }
      return cell;
    });
  });

  return {
    rows: rows as number,
    columns: columns as number,
    header_rows: headerRows as number,
    column_labels: normalizedColumnLabels,
    cells,
  };
}

export type NormalizedAnnotation = ReturnType<typeof validateAnnotation>;

export function validateAnnotation(value: unknown, completing: boolean) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    invalid("Annotation must be an object.");
  }
  const input = value as Record<string, unknown>;
  const documentType = optionalString(input.document_type, "Document type");
  if (documentType && !taxonomySets.documentTypes.has(documentType)) {
    invalid("Unknown physical document type.");
  }
  const variant = optionalString(input.document_variant, "Document variant");
  if (variant) {
    const contract = taxonomySets.variants.get(variant);
    if (!contract) invalid("Unknown physical document variant.");
    if (documentType && contract!.physical_document_type !== documentType) {
      invalid("Document variant does not belong to the selected type.");
    }
  }
  const contentProfile = optionalString(input.content_profile, "Content profile");
  if (contentProfile && !taxonomySets.contentProfiles.has(contentProfile)) {
    invalid("Unknown page content profile.");
  }
  const quality = input.image_quality ?? [];
  if (!Array.isArray(quality) || !quality.every((item) => typeof item === "string")) {
    invalid("Image quality must be an array.");
  }
  if (new Set(quality).size !== quality.length) invalid("Image quality values must be unique.");
  if (quality.some((item) => !taxonomySets.quality.has(item))) invalid("Unknown image quality flag.");
  if (quality.includes("clear") && quality.length > 1) {
    invalid("Clear cannot be combined with an image defect.");
  }
  const notes = input.notes ?? "";
  if (typeof notes !== "string" || notes.length > 4000) {
    invalid("Notes must contain at most 4,000 characters.");
  }
  const rawRegions = input.regions ?? [];
  if (!Array.isArray(rawRegions)) invalid("Regions must be an array.");
  const ids = new Set<string>();
  const orders = new Set<number>();
  const regions = rawRegions.map((raw, offset) => {
    const number = offset + 1;
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) invalid(`Region ${number} must be an object.`);
    const region = raw as Record<string, unknown>;
    const id = region.id;
    if (typeof id !== "string" || !REGION_ID.test(id)) invalid(`Region ${number} has an invalid ID.`);
    if (ids.has(id)) invalid("Region IDs must be unique.");
    ids.add(id);
    const label = region.label;
    const labelContract = typeof label === "string" ? taxonomySets.regionLabels.get(label) : undefined;
    if (!labelContract) invalid(`Region ${number} has an unknown label.`);
    const structure = region.structure_role ?? "none";
    if (typeof structure !== "string" || !taxonomySets.structures.has(structure)) {
      invalid(`Region ${number} has an unknown structure role.`);
    }
    const legibility = region.legibility;
    if (typeof legibility !== "string" || !taxonomySets.legibility.has(legibility)) {
      invalid(`Region ${number} has unknown legibility.`);
    }
    const readingOrder = region.reading_order;
    if (!Number.isInteger(readingOrder) || (readingOrder as number) < 1) {
      invalid(`Region ${number} needs a positive reading order.`);
    }
    if (orders.has(readingOrder as number)) invalid("Reading orders must be unique.");
    orders.add(readingOrder as number);
    const geometry = Object.fromEntries(
      ["x", "y", "width", "height"].map((name) => {
        const item = region[name];
        if (typeof item !== "number" || !Number.isFinite(item)) invalid(`Region ${number} has invalid geometry.`);
        return [name, Math.round(item * 1_000_000) / 1_000_000];
      }),
    ) as Record<"x" | "y" | "width" | "height", number>;
    if (geometry.x < 0 || geometry.x >= 100 || geometry.y < 0 || geometry.y >= 100) {
      invalid(`Region ${number} starts outside the page.`);
    }
    if (geometry.width <= 0 || geometry.height <= 0) invalid(`Region ${number} has no area.`);
    if (geometry.x + geometry.width > 100.001 || geometry.y + geometry.height > 100.001) {
      invalid(`Region ${number} extends outside the page.`);
    }
    const transcription = region.transcription ?? "";
    if (typeof transcription !== "string" || transcription.length > 10_000) {
      invalid(`Region ${number} has invalid transcription.`);
    }
    const rawFieldCode = region.field_code ?? "";
    if (typeof rawFieldCode !== "string" || rawFieldCode.length > 100) {
      invalid(`Region ${number} has invalid field code.`);
    }
    const fieldCode = rawFieldCode.trim();
    if (fieldCode && !FIELD_CODE.test(fieldCode)) invalid("Field codes must use lowercase snake_case.");
    if (labelContract!.textual && legibility === "not_applicable") {
      invalid(`Region ${number} is text and needs legibility.`);
    }
    if (["illegible", "not_applicable"].includes(legibility) && transcription.trim()) {
      invalid(`Region ${number} cannot contain a transcription.`);
    }
    if (
      completing &&
      labelContract!.requires_transcription_when_readable &&
      ["legible", "partially_legible"].includes(legibility) &&
      !transcription.trim()
    ) {
      invalid(`Region ${number} needs exact transcription.`);
    }

    const tableData = structure === "table"
      ? normalizeTableData(region.table_data, number)
      : null;
    if (structure !== "table" && region.table_data !== null && region.table_data !== undefined) {
      invalid(`Region ${number} has table data but is not marked as a table.`);
    }
    if (completing && tableData && !tableData.cells.some((row) => row.some((cell) => cell.trim()))) {
      invalid(`Region ${number} table needs at least one transcribed cell.`);
    }

    return {
      id,
      label: label as string,
      structure_role: structure,
      legibility,
      reading_order: readingOrder as number,
      field_code: fieldCode || null,
      transcription,
      table_data: tableData,
      ...geometry,
    };
  });
  if (completing) {
    if (!documentType) invalid("Select the physical document type before completing.");
    if (!contentProfile) invalid("Select the page content profile before completing.");
    if (!NON_REGION_DOCUMENTS.has(documentType!) && regions.length === 0) {
      invalid("Clinical pages need at least one annotated region.");
    }
  }
  return {
    schema_version: ANNOTATION_SCHEMA,
    document_type: documentType,
    document_variant: variant,
    content_profile: contentProfile,
    image_quality: quality,
    notes,
    regions: regions.sort((a, b) => a.reading_order - b.reading_order),
  };
}
